controllers/user_controller.go

```diff
@@ -0,0 +1,252 @@
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+var apiGVStr = topologyv1alpha1.GroupVersion.String()
+
+const (
+	userFinalizer = "deletion.finalizers.users.rabbitmq.com"
+	ownerKey      = ".metadata.controller"
+	ownerKind     = "User"
+)
+
+// UserReconciler reconciles a User object
+type UserReconciler struct {
+	client.Client
+	Log      logr.Logger
+	Scheme   *runtime.Scheme
+	Recorder record.EventRecorder
+}
+
+// +kubebuilder:rbac:groups=rabbitmq.com,resources=users,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=rabbitmq.com,resources=users/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups="",resources=secrets,verbs=get;list;watch;create;update
+
+func (r *UserReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := ctrl.LoggerFrom(ctx)
+
+	user := &topologyv1alpha1.User{}
+	if err := r.Get(ctx, req.NamespacedName, user); err != nil {
+		return reconcile.Result{}, client.IgnoreNotFound(err)
+	}
+
+	rabbitClient, err := rabbitholeClient(ctx, r.Client, user.Spec.RabbitmqClusterReference)
+	if err != nil {
+		logger.Error(err, "Failed to generate http rabbitClient")
+		return reconcile.Result{}, err
+	}
+
+	// Check if the user has been marked for deletion
+	if !user.ObjectMeta.DeletionTimestamp.IsZero() {
+		logger.Info("Deleting")
+		return ctrl.Result{}, r.deleteUser(ctx, rabbitClient, user)
+	}
+
+	if err := r.addFinalizerIfNeeded(ctx, user); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	spec, err := json.Marshal(user.Spec)
+	if err != nil {
+		logger.Error(err, "Failed to marshal binding spec")
+	}
+
+	logger.Info("Start reconciling",
+		"spec", string(spec))
+
+	if err := r.declareCredentials(ctx, user); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if err := r.setUserStatus(ctx, user); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if err := r.declareUser(ctx, rabbitClient, user); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	logger.Info("Finished reconciling")
+
+	return ctrl.Result{}, nil
+}
+
+func (r *UserReconciler) declareCredentials(ctx context.Context, user *topologyv1alpha1.User) error {
+	logger := ctrl.LoggerFrom(ctx)
+
+	// TODO: If a user has provided a Secret containing the desired password, use it instead here.
+	password, err := internal.RandomEncodedString(24)
+	if err != nil {
+		msg := "failed to generate random password"
+		r.Recorder.Event(user, corev1.EventTypeWarning, "FailedDeclare", msg)
+		logger.Error(err, msg)
+		return err
+	}
+
+	credentialSecret := corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      user.ObjectMeta.Name + "-user-credentials",
+			Namespace: user.ObjectMeta.Namespace,
+		},
+		Type: corev1.SecretTypeOpaque,
+		Data: map[string][]byte{
+			"username": []byte(base64.StdEncoding.EncodeToString([]byte(user.Spec.Name))),
```

Comment:
I would add a comment explaining that keys username and password are conforming to the provisioned service binding spec, and they are not random keys we have chosen ourselves